### PR TITLE
Fix for ticket:6037

### PR DESCRIPTION
**Pull request by adrpo**
- use stopTime for initSample if stopTimeDefined is set to true
- set status to fmi2Error if doStep goes over the defined stopTime

### CODE REVIEW

File: OMCompiler/SimulationRuntime/fmi/export/openmodelica/fmu2_model_interface.c.inc [hunk 1]
```diff
@@ -647,8 +647,8 @@ fmi2Status fmi2ExitInitializationMode(fmi2Component c)
     }
   }
 
-  /* TODO: Simulation stop time is needed to calculate the sample events beforehand. */
-  initSample(comp->fmuData, comp->threadData, comp->fmuData->localData[0]->timeValue, 100 /*should be stopTime*/);
+  /* use defined stopTime, if stopTimeDefined is given to calculate the sample events beforehand. */
+  initSample(comp->fmuData, comp->threadData, comp->fmuData->localData[0]->timeValue, comp->stopTimeDefined ? comp->stopTime : 100);
```
Comment (clagms) [inline OMCompiler/SimulationRuntime/fmi/export/openmodelica/fmu2_model_interface.c.inc:651]: It's strange to be mixing integer numbers with floats... What is 100? Is that the number of slots to be used for allocation? Or the default simulation time?

Comment (adrpo) [reply OMCompiler/SimulationRuntime/fmi/export/openmodelica/fmu2_model_interface.c.inc:651]: The initiSample is defined as:
`void initSample(DATA *data, threadData_t *threadData, double start, double stop);`
so it doesn't really matter if you send it 100.0 or 100. 
That is some sort of default for stopTime, if stopTime is not defined.

File: OMCompiler/SimulationRuntime/fmi/export/openmodelica/fmu2_model_interface.c.inc [hunk 2]
```diff
@@ -647,8 +647,10 @@ fmi2Status fmi2ExitInitializationMode(fmi2Component c)
     }
   }
 
-  /* TODO: Simulation stop time is needed to calculate the sample events beforehand. */
-  initSample(comp->fmuData, comp->threadData, comp->fmuData->localData[0]->timeValue, 100 /*should be stopTime*/);
+  /* use defined stopTime, if stopTimeDefined is given to calculate the sample events beforehand.
+   * TODO: when stopTime is not defined we use an arbitrary constant 100.0, maybe issue a warning
+   */
+  initSample(comp->fmuData, comp->threadData, comp->fmuData->localData[0]->timeValue, comp->stopTimeDefined ? comp->stopTime : 100.0 /* default stopTime */);
```
Comment (lochel) [inline OMCompiler/SimulationRuntime/fmi/export/openmodelica/fmu2_model_interface.c.inc:653]: @adrpo Can we recalculate the events if we exceed `100.0`?

Comment (adrpo) [reply OMCompiler/SimulationRuntime/fmi/export/openmodelica/fmu2_model_interface.c.inc:653]: I wondered about that, I don't really know. It was 100 before which seems arbitrary.

Comment (lochel) [reply OMCompiler/SimulationRuntime/fmi/export/openmodelica/fmu2_model_interface.c.inc:653]: I know that we had the same odd behavior before. However, it is a major issue for certain relevant applications and people probably don't know about it. And I would rather like to get this fixed than to tell people about our arbitrary 100.

Comment (adrpo) [reply OMCompiler/SimulationRuntime/fmi/export/openmodelica/fmu2_model_interface.c.inc:653]: Agreed, but somebody else needs to fix it as I'm not sure how. I guess one should monitor if it goes about the default value and somehow re-allocate the data in initSample.

**Pull request merged**